### PR TITLE
Update class_maker scripts

### PR DESCRIPTION
**Pull request by martyngigg**
Fixes an issue with `class_maker` where if a `CMakeLists.txt` file contained more than 1 reference to a variable such as `SRC_FILES` then the tool would try to modify all instances and the file would be corrupted. The script now only modifies the first instance of `set ( SRC_FILES` etc found. In addition there is also a new `--no-cmake` argument to prevent any modifications to cmake files.

I have also taken the opportunity to modify the scripts to be compatible with Python 2 & 3.

**To test:**

Use `class_maker` to generate a file in the `Geometry` library. The files should be created and the modifications to the `CMakeLists.txt` should only touch the expected lines at the top and **not** modify lines referencing opencascade later in the file.

Try out the `--no-cmake` option to confirm no cmake files are touched.

No issue. Internal change.

_Does not need to be in the release notes._

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [x] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: buildconfig/class_maker.py [hunk 1]
```diff
@@ -78,13 +79,13 @@ def write_header(subproject, classname, filename, args):
...
+class MANTID_{}_DLL {}{} {{
+public:{}}};
 
-} // namespace %s
-} // namespace Mantid
+}} // namespace {}
+}} // namespace Mantid
 
-#endif /* %s */""" % (guard, guard, subproject,
+#endif /* {} */""".format(guard, guard, subproject,
```
Comment (soininen) [inline buildconfig/class_maker.py]: Here one could save a bit of repetition with regards to the format parameters if named arguments were used:
```python
"""#ifndef {guard}
#define {guard}
#include "Mantid{subproject}/Dll.Config.h"
# [...]
#endif /* {guard} */""".format(guard = guard, subproject = subproject)
```
 Might also make the template strings easier to comprehend.

File: buildconfig/class_maker.py [hunk 2]
```diff
@@ -141,29 +142,29 @@ def write_source(subproject, classname, filename, args):
...
-} // namespace Mantid
-""" % (
+    s = """#include "Mantid{}/{}{}.h"
+
+namespace Mantid {{
+namespace {} {{
+{}{}
+}} // namespace {}
+}} // namespace Mantid
+""".format(
```
Comment (soininen) [inline buildconfig/class_maker.py]: Also here formatting field names might be nice.

File: buildconfig/class_maker.py [hunk 3]
```diff
@@ -174,24 +175,24 @@ def write_source(subproject, classname, filename, args):
 #======================================================================
 def write_test(subproject, classname, filename, args):
     """Write a class test file"""
-    print "Writing test file to %s" % filename
+    print("Writing test file to",filename)
```
Comment (soininen) [inline buildconfig/class_maker.py]: Missing space before `filename`?

File: buildconfig/class_maker.py [hunk 4]
```diff
@@ -207,39 +208,39 @@ def write_test(subproject, classname, filename, args):
...
-  }
+  }}
 
 
-};
+}};
 
 
-#endif /* %s */""" % (
+#endif /* {} */""".format(
```
Comment (soininen) [inline buildconfig/class_maker.py]: Formatting field names?

File: buildconfig/cmakelists_utils.py [hunk 5]
```diff
@@ -28,27 +30,29 @@ def find_basedir(project, subproject):
...
     read all the lines to get all the files,
     add your new line,
     sort them,
-    rewrite. Yay!"""
-
+    rewrite. Only touches first section found to avoid messing up any other set
+    sections in the rest of the file
+    """
     search_for1 = "set ( %s" % cmake_tag
     search_for2 = "set (%s" % cmake_tag
```
Comment (soininen) [inline buildconfig/cmakelists_utils.py]: Shouldn't we use the new-style string formatting here as well?

File: buildconfig/delete_class.py [hunk 6]
```diff
@@ -37,8 +37,8 @@ def delete_all(subproject, classname, args):
     # Insert into the cmake list
     remove_from_cmake(subproject, classname, args, args.source_subfolder)
 
-    print "   Files were removed to Framework/%s/CMakeLists.txt !" % subproject
-    print
+    print("   Files were removed to Framework/%s/CMakeLists.txt !" % subproject)
+    print()
```
Comment (soininen) [inline buildconfig/delete_class.py]: New style string formatting?

File: buildconfig/move_class.py [hunk 7]
```diff
@@ -44,7 +44,7 @@ def move_one(subproject, classname, newproject, newclassname, oldfilename, newfi
         f = open(newfilename, 'w')
         f.write(text)
     except RuntimeError as err:
-        print err
+        print(err)
```
Comment (soininen) [inline buildconfig/move_class.py]: Should we write error to stderr?

File: buildconfig/move_class.py [hunk 8]
```diff
@@ -64,16 +64,16 @@ def move_all(subproject, classname, newproject, newclassname, args):
...
+        print("\nError! Header file %s already exists. Use --force to overwrite.\n" % newheaderfile)
         return
     if args.cpp and not overwrite and os.path.exists(newsourcefile):
-        print "\nError! Source file %s already exists. Use --force to overwrite.\n" % newsourcefile
+        print("\nError! Source file %s already exists. Use --force to overwrite.\n" % newsourcefile)
         return
     if args.test and not overwrite and os.path.exists(newtestfile):
-        print "\nError! Test file %s already exists. Use --force to overwrite.\n" % newtestfile
+        print("\nError! Test file %s already exists. Use --force to overwrite.\n" % newtestfile)
         return
```
Comment (soininen) [inline buildconfig/move_class.py]: New style string formatting?

**Comment by soininen** [comment]
I tried to test this by running commands like

```
python2 buildconfig/class_maker.py Geometry ClassMakerTestGeometry
```

and

```
python2 buildconfig/class_maker.py --subfolder Rendering Geometry OCRenderingStuff1
```

on the master branch, but didn't see any problems with the `CMakeLists.txt` files. Could you perhaps elaborate how to reproduce the error cases? The changes to `redo_cmake_section()` seems to be doing what is expected.

Anyway, after the changes, the script runs on Python 3.4.3 and both python 2.7.6 and 3.4.3 generate the same files and changes. The small enhancements to the generated files compared to previous behavior are welcome, I think.

**Comment by martyngigg** [comment]
Thanks for checking this over. It's a good point about the formatting specifiers with keywords, I was being lazy by not updating them. I'll take a look at that.

It looks like there are no files left on `master` that exhibit the original error. The problem with the `Geometry` file was resolved by https://github.com/mantidproject/mantid/commit/b8c570c0003bfbc1c33397272b5a96403000a136. I have a branch where I have a similar issue in a different file so I was prompted to solve it for that. If you revert back to the way the `CMakeLists.txt` was in that diff then you should see that the new script still works.

**Comment by soininen** [comment]
I run the script before and after the changes in the `Geometry`'s `CMakeLists.txt` and indeed, there was some unexpected modifications to the `CMakelists.txt` file before the fixes of this PR. This also brings us closer to Python 3 compatibility. Great job! :shipit:

**Pull request merged**